Add two new guides for Yjs and Tiptap (#2757)
liveblocks/liveblocks: guides/pages/tiptap-best-practices-and-tips.mdx

---
meta:
  title: "Tiptap best practices and tips"
  description:
    "Learn best practices and common pitfalls to avoid when working with Tiptap
    and Liveblocks"
---

This guide covers best practices and common pitfalls to avoid when working with
Tiptap and Liveblocks.

<Banner title="Tiptap is backed by Yjs" type="note">

Tiptap uses [Yjs](https://docs.yjs.dev/) for synchronization. Make sure to also
check out our [Yjs best practices and tips](/guides/yjs-best-practices-and-tips)
guide for additional recommendations.

</Banner>

## Always include StarterKit or Doc, Paragraph extensions

The Tiptap
[StarterKit](https://tiptap.dev/docs/editor/extensions/functionality/starterkit)
extension is **required** for Tiptap to function properly. It provides essential
node types that every Tiptap editor needs, specifically the `Doc` and
`Paragraph` nodes. Alternatively you can include only the Doc and Paragraph

> extensions.

Reviewer: ```suggestion
`Paragraph` nodes. Alternatively you can include only the [Doc](https://tiptap.dev/docs/editor/extensions/nodes/document) and [Paragraph](https://tiptap.dev/docs/editor/extensions/nodes/paragraph)
extensions.
```

- The schema changes between different versions of your application
- Users collaborate on documents with different editor configurations

### How to enable content validation

Always enable content validation by setting `enableContentCheck: true` and
implementing an `onContentError` handler:

```tsx
import { useEditor } from "@tiptap/react";
import StarterKit from "@tiptap/starter-kit";

const editor = useEditor({
  extensions: [
    StarterKit.configure({
      history: false, // Required - disable history for Liveblocks
    }),
  ],
  enableContentCheck: true,
  onContentError: ({ editor, error, disableCollaboration }) => {
    // Disable collaboration to prevent data corruption
    disableCollaboration();

    // Make the editor read-only to prevent further changes
    editor.setEditable(false, false);

    // Log the error for debugging
    console.error("Content validation error:", error);

    // Notify the user that there's an issue
    alert(
      "There was an error loading this document. The content may be incompatible with the current editor version. The document has been made read-only to prevent data loss."
    );
  },

> });

Reviewer: ```suggestion
import { useEditor } from "@tiptap/react";
import StarterKit from "@tiptap/starter-kit";

const editor = useEditor({
  extensions: [
    StarterKit.configure({
      // Required, Liveblocks extension handles its own history
      undoRedo: false,
    }),
  ],
  // +++
  enableContentCheck: true,
  onContentError: ({ editor, error, disableCollaboration }) => {
    // Disable collaboration to prevent data corruption
    disableCollaboration();

    // Make the editor read-only to prevent further changes
    editor.setEditable(false, false);

    // Log the error for debugging
    console.error("Content validation error:", error);

    // Notify the user that there's an issue
    alert(
      "There was an error loading this document. The content may be incompatible with the current editor version. The document has been made read-only to prevent data loss."
    );
  },
  // +++
});
```

```

This ensures that:

- Invalid content is detected early
- Collaboration is disabled to prevent data corruption
- The editor is made read-only to prevent further changes
- The issue is logged for debugging
- Users are notified of the problem

## Use initialContent instead of content

When setting default content for your Tiptap editor, always use `initialContent`

> instead of `content`. Using `content` will cause the content to be duplicated

Reviewer: ```suggestion
instead of `content` on `useEditor`. Using `content` will cause the content to be appended to the document
```

```tsx
// ❌ AVOID: This will duplicate content on every page load
const editor = useEditor({
  extensions: [
    StarterKit.configure({
      history: false, // Required - disable history for Liveblocks
    }),
    liveblocks,
  ],
  content: "<p>Default text</p>", // This gets added every time!
});
```

### Use initialContent instead

The `initialContent` option sets a flag internally and only sets the content the
**very first time** the document is empty. This prevents duplication:

```tsx
// ✅ CORRECT: This only sets content once
import { useEditor } from "@tiptap/react";
import { useLiveblocksExtension } from "@liveblocks/react-tiptap";
import StarterKit from "@tiptap/starter-kit";

function Editor() {
  const liveblocks = useLiveblocksExtension();

  const editor = useEditor({
    extensions: [
      StarterKit.configure({
        history: false, // Required - disable history for Liveblocks
      }),
      liveblocks,
    ],
    initialContent: "<p>Default text</p>", // Only set on first load
  });

  return <EditorContent editor={editor} />;
}
```

<Banner title="Not compatible with the field option" type="warning">

> The `initialContent` option does not currently work when using the `field`

Reviewer: ```suggestion
The [`initialContent`](/docs/api-reference/liveblocks-react-tiptap#Setting-initial-content) option does not currently work when using the [`field`](/docs/api-reference/liveblocks-react-tiptap#Multiple-editors)
```

option for multiple editors. If you need to set initial content for multiple
editors, you'll need to implement your own logic to check if the document is
empty before setting content.

</Banner>

## Support multiple editors with the field option

If you want to display multiple Tiptap editors on the same page, use the `field`
option with a unique identifier for each editor. This ensures that each editor
synchronizes to its own section of the Yjs document.

```tsx
import { useEditor } from "@tiptap/react";
import { useLiveblocksExtension } from "@liveblocks/react-tiptap";
import StarterKit from "@tiptap/starter-kit";

function EditorOne() {
  const liveblocks = useLiveblocksExtension({
    field: "editor-1", // Unique identifier for this editor
  });

  const editor = useEditor({
    extensions: [
      StarterKit.configure({

>         history: false, // Required - disable history for Liveblocks

Reviewer: ```suggestion
        // Required, Liveblocks extension handles its own history
        undoRedo: false,
```

      }),
      liveblocks,
    ],
  });

  return <EditorContent editor={editor} />;
}

function EditorTwo() {
  const liveblocks = useLiveblocksExtension({
    field: "editor-2", // Different unique identifier
  });

  const editor = useEditor({
    extensions: [
      StarterKit.configure({
        history: false, // Required - disable history for Liveblocks
      }),
      liveblocks,
    ],
  });

  return <EditorContent editor={editor} />;
}
```

Without the `field` option, both editors would synchronize to the same location
in the document, causing conflicts and data loss.

[Learn more about multiple editors →](/docs/api-reference/liveblocks-react-tiptap#Multiple-editors)

## Never use extensions with binary data

Never use Tiptap extensions that store binary data (such as base64-encoded
images) directly in the document. Binary data will be synchronized across all

> clients and can fill up your room extremely quickly, leading to:

Reviewer: ```suggestion
clients and can fill up your Liveblocks room extremely quickly, leading to:
```

- Increased bandwidth usage
- Slower synchronization
- Higher storage costs
- Potential rate limiting

### Common pitfall: Image extension

The official Tiptap
[Image extension](https://tiptap.dev/docs/editor/extensions/nodes/image) has an
`allowBase64` option. This option is **defaulted to `false`**, and it should
**never be set to `true`**.

```tsx
// ❌ NEVER DO THIS
import Image from "@tiptap/extension-image";

const editor = useEditor({
  extensions: [
    StarterKit.configure({

>       history: false, // Required - disable history for Liveblocks

Reviewer: ```suggestion
      // Required, Liveblocks extension handles its own history
      undoRedo: false,
```

    }),
    Image.configure({
      allowBase64: true, // This will cause problems!
    }),
  ],
});
```

### Recommended approach

Instead of storing images as base64 in the document:

1. Upload images to a file storage service (e.g., AWS S3, Cloudflare R2, Vercel
   Blob)
2. Store only the URL in the document
3. Reference the URL in your image nodes

```tsx
// ✅ CORRECT: Store only URLs
import Image from "@tiptap/extension-image";

const editor = useEditor({
  extensions: [
    StarterKit.configure({
      history: false, // Required - disable history for Liveblocks
    }),
    Image.configure({
      allowBase64: false, // Default, but explicit is better
    }),
  ],
});

// When adding an image
editor.commands.setImage({
  src: "https://your-storage.com/images/photo.jpg", // URL only, no base64

> });

Reviewer: ```suggestion
// +++
editor.commands.setImage({
  // URL only, no base64
  src: "https://your-storage.com/images/photo.jpg", 
});
// +++
```
